compiler/qsc_linter/src/linter/ast.rs

```diff
@@ -0,0 +1,190 @@
+use qsc_ast::{
+    ast::{
+        Attr, Block, CallableDecl, Expr, FunctorExpr, Ident, Item, Namespace, Package, Pat, Path,
+        QubitInit, SpecDecl, Stmt, TopLevelNode, Ty, TyDef, Visibility,
+    },
+    visit::{self, Visitor},
+};
+
+use crate::{
+    lints::ast::{DivisionByZero, DoubleParens},
+    Lint,
+};
+
+/// The entry point to the AST linter. It takes a [`qsc_ast::ast::Package`]
+/// as input and outputs a [`Vec<Lint>`](Lint).
+#[must_use]
+pub fn run_ast_lints(package: &qsc_ast::ast::Package) -> Vec<Lint> {
+    let mut lints = CombinedAstLints::default();
+
+    for node in package.nodes.iter() {
+        match node {
+            TopLevelNode::Namespace(namespace) => {
+                lints.visit_namespace(namespace);
+            }
+            TopLevelNode::Stmt(stmt) => {
+                lints.visit_stmt(stmt);
+            }
+        }
+    }
+
+    lints.buffer
+}
+
+/// Combined AST lints for speed. This combined lint allow us to
+/// evaluate all the lints in a single AST pass, instead of doing
+/// an individual pass for each lint in the linter.
+#[derive(Default)]
+pub(crate) struct CombinedAstLints {
+    buffer: Vec<Lint>,
+}
+
+#[allow(clippy::unused_self)]
+impl CombinedAstLints {
+    fn check_package(&self, _package: &Package) {}
+
+    fn check_namespace(&self, _namespace: &Namespace) {}
+
+    fn check_item(&self, _item: &Item) {}
+
+    fn check_attr(&self, _attr: &Attr) {}
+
+    fn check_visibility(&self, _visibility: &Visibility) {}
+
+    fn check_ty_def(&self, _def: &TyDef) {}
+
+    fn check_callable_decl(&self, _decl: &CallableDecl) {}
+
+    fn check_spec_decl(&self, _decl: &SpecDecl) {}
+
+    fn check_functor_expr(&self, _expr: &FunctorExpr) {}
+
+    fn check_ty(&self, _ty: &Ty) {}
+
+    fn check_block(&self, _block: &Block) {}
+
+    fn check_stmt(&self, _stmt: &Stmt) {}
+
+    fn check_expr(&mut self, expr: &Expr) {
```

Comment:
I could put any function or logic in these that I want, right? In general each lint listed here is a function that checks the given ast node, and puts a Lint in the self.buffer if needed, right?